api/v2/db/ldap/connection.js

```
@@ -0,0 +1,24 @@
+const config = require('config');
+const ldap = require('ldapjs');
+
+const { url } = config.get('dataSources.ldap');
+
+/**
+ * @summary Get an ldap connection
+ * @function
+ * @returns {ldapjs.client} ldap client connection object
+ */
+const getClient = () => ldap.createClient({ url });
+
+/**
+ * @summary Validate ldap connection and throw an error if invalid
+ * @function
+ * @throws Throws an error if unable to connect or search ldap
+ */
+const validateLdap = async () => {
+  ldap.createClient({ url }).on('error', () => {
+    throw new Error('Error connecting to ldap');
```

It looks like `createClient` will only emit an `error` event if you pass in `bindDN` and `bindCredentials` which will cause an implicit bind operation https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L346. Since that's not happening, this shouldn't be triggered. Also exceptions may be thrown here if the url is invalid for example so the call here should be wrapped in a try-catch block if you want to throw a different error.

---

`createClient` could also emit an `error` event on a failed attempt at reconnecting:
https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L1228
https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L1290

```
Error: Error connecting to ldap
    at Client.<anonymous> (/home/ian/directory-api/api/v2/db/ldap/connection.js:20:11)
    at Client.emit (events.js:200:13)
    at Backoff.<anonymous> (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1228:12)
    at Backoff.emit (events.js:200:13)
    at Backoff.backoff (/home/ian/directory-api/node_modules/backoff/lib/backoff.js:41:14)
    at /home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1214:15
    at f (/home/ian/directory-api/node_modules/once/once.js:25:25)
    at TLSSocket.onResult (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1016:7)
    at Object.onceWrapper (events.js:288:20)
    at TLSSocket.emit (events.js:200:13)
```

This is the error that is omitted when an invalid url is used in the configuration. When wrapping with a try / catch block, the exception is not actually caught.
```
Error: getaddrinfo ENOTFOUND ldap.fsdafsdonid.orst.edu
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
Emitted 'error' event at:
    at Backoff.<anonymous> (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1228:12)
    at Backoff.emit (events.js:200:13)
    at Backoff.backoff (/home/ian/directory-api/node_modules/backoff/lib/backoff.js:41:14)
    at /home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1214:15
    at f (/home/ian/directory-api/node_modules/once/once.js:25:25)
    at TLSSocket.onResult (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1016:7)
    at Object.onceWrapper (events.js:288:20)
    at TLSSocket.emit (events.js:200:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3) {
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'ldap.fsdafsdonid.orst.edu'
}
events.js:177
      throw er; // Unhandled 'error' event
```

---

Hm, I tried the url `ldap.fsdafsdonid.orst.edu` and got this exception:
```
TypeError: ldap.fsdafsdonid.orst.edu is an invalid LDAP url (protocol)
    at Object.parse (/Users/ian/github/directory-api/node_modules/ldapjs/lib/url.js:16:13)
    at new Client (/Users/ian/github/directory-api/node_modules/ldapjs/lib/client/client.js:310:16)
    at Object.createClient (/Users/ian/github/directory-api/node_modules/ldapjs/lib/client/index.js:54:12)
    at Object.validateLdap [as ldap] (/Users/ian/github/directory-api/api/v2/db/ldap/connection.js:19:23)
    at _.each (/Users/ian/github/directory-api/utils/validate-data-source.js:36:40)
    at arrayEach (/Users/ian/github/directory-api/node_modules/lodash/lodash.js:516:11)
    at Function.forEach (/Users/ian/github/directory-api/node_modules/lodash/lodash.js:9344:14)
    at validateDataSource (/Users/ian/github/directory-api/utils/validate-data-source.js:34:5)
    at Object.<anonymous> (/Users/ian/github/directory-api/app.js:22:1)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
```
I'm not sure why we got different errors here but there does appear to be at least one case where an exception may be thrown. It looks like there's documentation on this too: https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L300